### ISSUE TITLE
Mempool saving feature between restarts

### ISSUE THREAD

**Issue by CryptoDwarf** [bitcoinxt/bitcoinxt#195]
Bitcoin core 0.14 has a mempool saving feature that saves the mempool states between restarts. This is a very useful feature for people who multiboot between O/S's. (and also in the event of a system crash for miners.) I think this would be useful to add to bitcoin XT.
I am not familiar with this part of the code, but I would look to implement the transaction pool in a virtual memory segment using mmap(), msync(), munmap() if possible. This would mean that all transactions are immediately loaded into virtual memory on start up when the memory segment is mapped onto.

**Comment by CryptoDwarf** [comment]
I have opened similar issues on Classic and BU:

https://github.com/bitcoinclassic/bitcoinclassic/issues/251
https://github.com/BitcoinUnlimited/BitcoinUnlimited/issues/438

So far zander has provided some insights and thoughts on this issue, which are certainly worth considering.

**Comment by dagurval** [comment]
What's a good use case for this? I don't think recovery after system reboot is a good one.

If there is a good use case, how about seeding our mempool with other nodes mempool instead? It's possible to send a "mempool" request to other nodes to retrieve the contents of their pool.

**Comment by CryptoDwarf** [comment]
Yes, if you read the bitcoin classic issue, I came to the conclusion that synchronising the transaction pool from other peers on startup would be a better solution instead. I think the reason to do this is to prevent a possible double spending path through 'part-time' nodes which do not run 24x7, when the transaction may get rejected by the mining pools. That is quite important if zero confirmation economic risk is to have any meaning.
Makes me wonder why core implemented this in this fashion, especially if the mempool request already exists?

**Comment by dagurval** [comment]
Downside of requesting the mempool from other nodes is that it relatively bandwidth intensive.

Only use case I can think of is for better thin block performance at start up - if we plan on keeping the 1MB limit. With the limit transactions may stay unconfirmed for some hours. But we don't plan on keeping the limit :-).

**Comment by CryptoDwarf** [comment]
Zero confirmation economic risk is a whole different ball game with full blocks, as it changes from the disadvantageous double spend attempt to one where the transaction may not get confirmed at all and is dropped from forgetful mempools!
Even with bigger blocksizes, there is still potential for peak mempool flood activity, where the miners may not be able to keep up with demand quick enough for a period of time. So I still wonder it such a feature may still be desirable.

**Comment by CryptoDwarf** [comment]
gmaxwell as outlined the idea behind this feature.

https://bitcointalk.org/index.php?topic=1864311.0

**Comment by dagurval** [comment]
How about a more advanced "mempool" network request?

`mempool(filter, flags)`

where `filter` contains the transactions _you already have_. What you want to receive back is a list of transactions that **don't** match the filter. This is the opposite of what the current mempool command does. This is similar to `get_xthin` request.

By default a node receiving this request could reply with `<= current block size limit` amount transactions. That's currently 8MB on the Cash network. The reply should be the transactions deemed most important - for example the most fee paying transactions. The behaviour can be modified with flags.

This should add little or no bandwidth cost. You'll only receive transactions you don't have. The transactions are likely to be included in the next block. Thus the bandwidth used is returned next time we request a thin block.

Nodes could issue this mempool request when it is fully synced. Possibly periodically and when connecting to a new node.

Pros:
- New node will immediately be able to mine transactions.
- Next block will propagate faster.
- Reduced load on network when next block arrives (when the network is most busy)

Cons:
- Node receiving the requests gets to know your mempool. It can send you a double spending transaction and that you'd most likely accept it.
- Network works fine without this feature. Is it worth it?
